### PR TITLE
Add HttpSession

### PR DESCRIPTION
**Pull request by Geas77**

### CODE REVIEW

File: src/HttpSession.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+public:
+
+	HttpSession(const int &socket);
+	HttpSession(const HttpSession &) = delete;
+	/**@breif Run socketSession after initial.*/
+	void run();
+	/** @breif Receive stream package.
+	 *	@param msg Stream package's content.
+	 */
+	void receiveMessage(const std::string &msg);
```
Comment (john0312) [inline src/HttpSession.h:27]: Who calls this? If it's only used internally, then please make it private.

File: src/HttpSession.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	/** @breif Handle HttpRequest.
+	 *	@param request Decoded HttpRequest.
+	 */
+	void handleRequest(const HttpUtility::Request &request);
```
Comment (john0312) [inline src/HttpSession.h:31]: Is this used only internally?

**Comment by NTUST-CS-PSSE** [comment]
Hello @kuojohnny , please take a look at this commit/PR. Thank you.

**Comment by john0312** [comment]
Hello @kuojohnny , please take a look at this commit/PR. Thank you.

File: src/HttpSession.cpp [hunk 3]
```diff
@@ -0,0 +1 @@
+#include "HttpSession.h"
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.cpp]: Done

File: src/HttpSession.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2021, Group5 Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#include<iostream>
+#include<string>
+// SocketSession. 
+#include "SocketSession.h"
+#include "HttpUtility.h"
+class SocketSession;
+
+/**	@breif Handle the incoming http request. */
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+class HttpSession{
+private:
+	/**	@breif socket id.*/
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2021, Group5 Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+#include<iostream>
+#include<string>
+// SocketSession.
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+class SocketSession;
+
+/**	@breif Handle the incoming http request. */
+class HttpSession{
+private:
+	/**	@breif socket id.*/
+	int socket_;
+public:
+
+	HttpSession(const int &socket);
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	HttpSession(const HttpSession &) = delete;
```
[the remark this answers is not in the document]
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h (hunk 1 above)
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h (hunk 2 above)
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "SocketSession.h"
+#include "HttpUtility.h"
+class SocketSession;
+
+/** @brief Handle the incoming http request. */
+class HttpSession {
+private:
+    /** @brief socket fd.*/
+    int socket_;
```
Comment (john0312) [inline src/HttpSession.h:18]: Please use asio type.

File: src/HttpSession.h [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    /** @brief Receive stream package.
+     *  @param msg Stream package's content.
+     */
+    void receiveMessage(const std::string &msg);
+    /** @brief Handle HttpRequest.
+     *  @param request Decoded HttpRequest.
+     */
+    void handleRequest(const HttpUtility::Request &request);
+    /** @brief Pakage Response and Send.
```
Comment (john0312) [inline src/HttpSession.h:27]: Package

File: src/HttpSession.h [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+#include "SocketSession.h"
+#include "HttpUtility.h"
+class SocketSession;
+
+/** @brief Handle the incoming http request. */
+class HttpSession {
+private:
+    /** @brief socket fd.*/
+    int socket_;
+    /** @brief Receive stream package.
```
Comment (john0312) [inline src/HttpSession.h:19]: Please check the asio callback type.
Please comment that this is a callback for read_async() in asio.

File: src/HttpSession.h [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+     *  @param msg Stream package's content.
+     */
+    void receiveMessage(const std::string &msg);
+    /** @brief Handle HttpRequest.
+     *  @param request Decoded HttpRequest.
+     */
+    void handleRequest(const HttpUtility::Request &request);
```
Comment (john0312) [inline src/HttpSession.h:26]: Please add comment on this is called by receiveMessage() after the entire http request has been received.

File: src/HttpSession.h [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    /** @brief Pakage Response and Send.
+     *  @param response Decoded HttpResponse.
+     */
+    void sendMessage(const HttpUtility::Response &response);
+    /** @brief Decide  upgrade or not. */
```
Comment (john0312) [inline src/HttpSession.h:31]: "Return true if this is upgraded to websocket."

File: src/HttpSession.h [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    bool isUpgrade();
+
+public:
+
+    explicit HttpSession(const int &socket);
+    HttpSession(const HttpSession &) = delete;
```
Comment (john0312) [inline src/HttpSession.h:37]: Please group with copy operator and comment that this class is not copyable.

File: src/HttpSession.h [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+    /** @brief Run socketSession after initial.*/
+    void run();
+
+    /** @brief Close webSocket session.*/
```
Comment (john0312) [inline src/HttpSession.h:42]: Http Socket?

File: src/HttpSession.h (hunk 11 above)
Comment (Geas77) [reply src/HttpSession.h]: Done

File: src/HttpSession.h (hunk 14 above)
Comment (Geas77) [reply src/HttpSession.h:29]: Done

**Review by john0312 — APPROVED** [review]
Approved assuming comments are addressed.

**Pull request merged**